explainer/index.bs

```diff
@@ -170,22 +174,246 @@ Of course, it becomes a validation to use the object after the call to `.destroy
+    Handling device/adapter loss, either by restoring WebGPU or by fallback to non-WebGPU content.
+- *Fallible Allocation*:
+    Making fallible GPU-memory resource allocations (detecting out-of-memory conditions).
+- *Fallible Validation*:
+    Checking success of WebGPU calls, for applications' unit/integration testing, WebGPU
+    conformance testing, or detecting errors in data-driven applications (e.g. loading glTF
+    models that may exceed device limits).
+- *Telemetry*:
+    Collecting error logs in deployment, for bug reporting and telemetry.
+
+The following sections go into more details on these cases and how they are solved.
+
+#### Debugging #### {#errors-cases-debugging}
+
+**Solution:** Dev Tools.
+
+Implementations should provide a way to enable synchronous validation,
+for example via a "break on WebGPU error" option in the developer tools.
+
+This can be achieved with a content-process-gpu-process round-trip in every validated WebGPU
+call, though in practice this would be very slow.
+It can be optimized by running a second, approximated mirror of the validation steps in the
+content process (it will not always have the same results since it cannot immediately know about
+out-of-memory errors).
+
+#### Fatal Errors: Adapter and Device Loss #### {#errors-cases-fatalerrors}
+
+**Solution:** [[#device-loss]].
+
+#### Fallible Allocation, Fallible Validation, and Telemetry #### {#errors-cases-other}
+
+**Solution:** *Error Scopes*.
+
+For important context, see [[#invalid-and-destroyed]]. In particular, all errors (validation and
+out-of-memory) are detected asynchronously, in a remote process.
+In the WebGPU spec, we refer to the thread of work for each WebGPU device as its "device timeline".
+
+As such, applications need a way to instruct the device timeline on what to do with any errors
+that occur. To solve this, WebGPU uses *Error Scopes*.
+
+### Error Scopes ### {#errors-errorscopes}
+
+Each device<sup>1</sup> maintains a persistent "error scope" stack state.
+Initially, the device's error scope stack is empty.
+`GPUDevice.pushErrorScope('validation')` or `GPUDevice.pushErrorScope('out-of-memory')`
+begins an error scope and pushes it onto the stack.
+
+`GPUDevice.popErrorScope()` ends an error scope, popping it from the stack and returning a
+`Promise<GPUError?>`, which resolves once all enclosed fallible operations have completed and
+reported back.
+It resolves to `null` if no errors were captured, and otherwise resolves to an object describing
+the first error that was captured by the scope - either a `GPUValidationError` or a
+`GPUOutOfMemoryError`.
+
+An error scope captures an error if its filter (`'validation'` or `'out-of-memory'`) matches the
+type of the error.
+This filter mechanism avoids, e.g., accidentally silencing validation errors when trying to
+implement fallible allocation.
+
+Any device-timeline error from an operation is passed to the top-most error scope on the stack at
+the time it was issued.
```

Comment:
@Kangz:
> Maybe explain that this we done for improved composability.

Not sure what you meant by this, could you explain a bit more?

Reply:
We should describe the reason we have a stack of error scopes and not just a single-level error scope (or something like async glGetError). It is to improve the composability of the application. For example a middleware can choose to catch OOM errors on an allocation while the rest of the application doesn't know about it. It makes the error API more stateless by making correctly written code (that balance error scopes) blind to whether it is already in error scopes or not.